lib/ansible/utils/collection_loader.py

```diff
@@ -0,0 +1,305 @@
+        if parent_pkg_name and not parent_pkg:
+            raise ImportError('parent package {0} not found'.format(parent_pkg_name))
+
+        # are we at or below the collection level? eg a.mynamespace.mycollection.something.else
+        # if so, we don't want distributed namespace behavior; first mynamespace.mycollection on the path is where
+        # we'll load everything from (ie, don't fall back to another mynamespace.mycollection lower on the path)
+        sub_collection = fullname.count('.') > 1
+
+        synpkg_def = _SYNTHETIC_PACKAGES.get(fullname)
+
+        # FIXME: collapse as much of this back to on-demand as possible (maybe stub packages that get replaced when actually loaded?)
+        if synpkg_def:
+            pkg_type = synpkg_def.get('type')
+            if not pkg_type:
+                raise KeyError('invalid synthetic package type (no package "type" specified)')
+            if pkg_type == 'map':
+                map_package = synpkg_def.get('map')
+
+                if not map_package:
+                    raise KeyError('invalid synthetic map package definition (no target "map" defined)')
+                mod = import_module(map_package)
+
+                sys.modules[fullname] = mod
+
+                return mod
+            elif pkg_type == 'flatmap':
+                raise NotImplementedError()
+            elif pkg_type == 'pkg_only':
+                newmod = ModuleType(fullname)
+                newmod.__package__ = fullname
+                newmod.__file__ = '<ansible_synthetic_collection_package>'
+                newmod.__loader__ = self
+                newmod.__path__ = []
+
+                sys.modules[fullname] = newmod
+
+                return newmod
+
+        if not parent_pkg:  # top-level package, look for NS subpackages on all collection paths
+            package_paths = [self._extend_path_with_ns(p, fullname) for p in self._collection_paths]
+        else:  # subpackage; search in all subpaths (we'll limit later inside a collection)
+            package_paths = [self._extend_path_with_ns(p, fullname) for p in parent_pkg.__path__]
+
+        for candidate_child_path in package_paths:
+            source = None
+            is_package = True
+            location = None
+            # check for implicit sub-package first
+            if os.path.isdir(candidate_child_path):
+                # Py3.x implicit namespace packages don't have a file location, so they don't support get_data
+                # (which assumes the parent dir or that the loader has an internal mapping); so we have to provide
+                # a bogus leaf file on the __file__ attribute for pkgutil.get_data to strip off
+                location = os.path.join(candidate_child_path, '__synthetic__')
+            else:
+                for source_path in [os.path.join(candidate_child_path, '__init__.py'),
+                                    candidate_child_path + '.py']:
+                    if os.path.isfile(source_path):
```

Comment:
could save some nesting
```suggestion
                    if not os.path.isfile(source_path):
                        continue
```